### PR TITLE
Moving synthetic materialization generation to not be passed on proxied state

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This PR changes that way that we determine whether or not to generate asset materializations for completed tasks.

Previously were we relying on state fetched at code location load time to make this determination. Unfortunately, this meant that as that state changed and the code location was not reloaded, things would get out of sync and we would create phantom materializations.

This relies on a new scheme that uses dagster runs as the source of truth.

When the sensor consumes an airflow dag run, it finds all the task instances that have successfully completed.  It also queries the dagster metadata store for all dagster runs tagged with that airflow dag run's ID, which was attached by the proxied operator.

If, for a given completed, successful task_id, there exists a Dagster run, we know that there has been a best effort to materialize the associated asset via a proxy operation. If the dagster run does not, then the execution happened within Airflow, and we should synthesize a materialization and attach airflow metadata information to it.

## How I Tested These Changes

Manual testing with "kitchen-sink" and automated testing.

Testing I would say is relatively weak here as we don't do this at massive scale and for a wide swath of use cases where there is weird state in the Airflow instance.

## Changelog

NOCHANGELOG

### CODE REVIEW

**Comment by schrockn** [comment]
* **#24981** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24981?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @schrockn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/constants.py [hunk 1]
```diff
@@ -3,4 +3,6 @@
 TASK_MAPPING_METADATA_KEY = "dagster-airlift/task_mapping"
 # This represents the timestamp used in ordering the materializatons.
 EFFECTIVE_TIMESTAMP_METADATA_KEY = "dagster-airlift/effective_timestamp"
-DAG_RUN_ID_TAG_KEY = "dagster-airlift/dag-run-id"
+DAG_RUN_ID_TAG_KEY = "dagster-airlift/airflow-dag-run-id"
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/dagster_airlift/constants.py:6]: should we have an additional prefix for "from airflow"

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py [hunk 2]
```diff
@@ -190,3 +194,70 @@ def materializations_and_requests_from_batch_iter(
...
+
+    check.invariant(
+        len({ti.task_id for ti in task_instances}) == len(task_instances),
+        "Assuming one task instance per task_id for now. Dynamic Airflow tasks not supported.",
+    )
+
+    # TODO: Need to properly paginate this.
+    dagster_runs = context.instance.get_runs(
+        filters=RunsFilter(tags={DAG_RUN_ID_TAG_KEY: dag_run.run_id}),
+        limit=1000,
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py:242]: looks paginated to me 😎

File: examples/experimental/dagster-airlift/examples/kitchen-sink/kitchen_sink_tests/integration_tests/test_e2e_mapped.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def poll_for_materialization(
     asset_key: AssetKey,
 ) -> EventLogEntry:
     start_time = get_current_datetime()
-    while get_current_datetime() - start_time < timedelta(seconds=30):
+    while get_current_datetime() - start_time < timedelta(seconds=300000):
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/examples/kitchen-sink/kitchen_sink_tests/integration_tests/test_e2e_mapped.py:21]: assuming this was necessary for testing but let's go back to a reasonable number here

**Review by dpeng817 — APPROVED** [review]

**Comment by dpeng817** [comment]
Let's get this in. Core logic mostly looks good, I'll add a battery of unit tests later and probably clean up the implementation a bit.

File: examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/core_tests/test_sensor.py [hunk 4]
```diff
@@ -30,7 +31,7 @@ def make_dag_key_str(dag_id: str) -> str:
     return make_dag_key(dag_id).to_user_string()
 
 
-def test_dag_and_task_metadata(init_load_context: None) -> None:
+def test_dag_and_task_metadata(init_load_context: None, instance: DagsterInstance) -> None:
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/core_tests/test_sensor.py:34]: where is this fixture even coming from

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py (hunk 2 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py:242]: Adding https://linear.app/dagster-labs/issue/FOU-444/make-sensor-work-with-an-airflow-dag-run-that-has-more-than-1000 to track

File: examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/core_tests/test_sensor.py (hunk 4 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/core_tests/test_sensor.py:34]: examples/experimental/dagster-airlift/dagster_airlift_tests/unit_tests/conftest.py

pytest is crazy

File: examples/experimental/dagster-airlift/examples/kitchen-sink/kitchen_sink_tests/integration_tests/test_e2e_mapped.py (hunk 3 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/examples/kitchen-sink/kitchen_sink_tests/integration_tests/test_e2e_mapped.py:21]: yes good catch thanks

File: examples/experimental/dagster-airlift/dagster_airlift/constants.py (hunk 1 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/constants.py:6]: well i thought the `airflow` prefix in front of `dag-run-id` et all implied that.

I think the pertinent information might be that it is the proxy that inserted it

**Comment by schrockn** [comment]
### Merge activity

* **Oct 2, 3:05 PM EDT**: @schrockn started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/24981).

**Pull request merged**